### ISSUE TITLE
Regexes in short-term archiver's get_file_date are too general

### ISSUE THREAD

**Issue by billsacks** [ESMCI/cime#2308]
`get_file_date` in `case_st_archive.py` defines these regular expressions:

```python
    # TODO: Add these to config_archive.xml, instead of here
    # Note these must be in order of most specific to least
    # so that lesser specificities aren't used to parse greater ones
    re_formats = [r"[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}_[0-9]{1,2}\.[0-9]{1,2}\.[0-9]{1,2}", # yyyy-mm-dd_hh.MM.ss
                  r"[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}[\-_][0-9]{1,5}",                     # yyyy-mm-dd_sssss
                  r"[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}",                                    # yyyy-mm-dd
                  r"[0-9]{4}[\-\.][0-9]{1,2}",                                          # yyyy-mm
    ]
```

But these regular expressions are too general. In particular, they don't define what should or shouldn't be at the starting and ending boundaries of these date strings.

I ran into trouble when the archiver tried to run on a case that had some numbers in the case name. My case name was `ERI_Lm24.f10_f10_musgs.I1850Clm50SpG.cheyenne_intel.cism-isostasy_period4.GC.0220-1646_i.ref1`. This generated this error when parsing a file name like `ERI_Lm24.f10_f10_musgs.I1850Clm50SpG.cheyenne_intel.cism-isostasy_period4.GC.0220-1646_i.ref1.mosart.h0.0001-01.nc`:

```
  File "/glade/p/work/sacks/cesm_code/cism-wrapper/cime/scripts/Tools/../../scripts/lib/CIME/case_st_archive.py", line 316, in _archive_history_files
    file_date = get_file_date(os.path.basename(histfile))
  File "/glade/p/work/sacks/cesm_code/cism-wrapper/cime/scripts/Tools/../../scripts/lib/CIME/case_st_archive.py", line 101, in get_file_date
    return datetime.datetime(year, month, day) + datetime.timedelta(seconds = second)
ValueError: month must be in 1..12
```

The problem is that the regex match for `r"[0-9]{4}[\-\.][0-9]{1,2}"` matched part of the case name (`0220-1646`), which it interpreted as year 0220 month 16 (ignoring the 46), rather than matching the actual date, `0001-01`.

I can see a few possible solutions; I'm not sure which of these are right things to do. It may be that we want to do more than one of these (i.e., either (1) or (2) and also (3)). Note that many of these solutions will require parenthesizing the portion of the regex that contains the date, then changing `date_str = match.group()` to `date_str = match.group(1)`.

1. Is it always the case that the date appears at the end of the file name, just before the `.nc`? If so, then `\.nc` should be added to the end of these regexes.

2. If we can't rely on the date always appearing just before the `.nc`, but we can rely on it being the last digits that appear in the file name, then we could find the last match rather than the first, e.g., using `findall` (though note that `findall` gives non-overlapping matches, so it would probably be possible to construct edge cases where this fails).

3. In my case, part of the problem was that the regexes pick up matches where there are extra digits before and after the ones it's trying to match. e.g., the match for `r"[0-9]{4}[\-\.][0-9]{1,2}"` is trying to just match one or two characters after the dash, but it matches the first seven characters of `0220-1646`. This could be addressed by adding negated character classes at the beginning and end of each regex, to make sure they match only the exact number of expected digits. This solution alone would be an improvement, though not as good as (1), because it would still break if a case name had something like `0220-16`.

**Comment by billsacks** [comment]
We discussed this on Wednesday's cime telecon. @mfdeakin wasn't sure if we can assume that the date always comes directly before the `.nc`. I think it would be worth looking into this a bit: If possible, it would be really great if we could make these regexes more robust by having a rule that the date always needs to appear directly before the `.nc`.

I confirmed that this is the case for CISM. I'll ask other CESM liaisons to confirm this for their components. @mfdeakin or someone else: can you check on this for E3SM?

**Comment by gold2718** [comment]
Both CAM and the  E3SM atmosphere model have a namelist item called `hfilename_spec` which allows the user to make the filename pattern be whatever they want. I have no idea if anyone uses this feature. There is also a similar `sathist_hfilename_spec` for satellite history files.

The restart filename patterns are hardcoded in both models and have the date just before the `.nc`.

**Comment by billsacks** [comment]
@bertinia reminded me of http://www.cesm.ucar.edu/models/cesm2.0/cesm/filename_conventions_cesm.html . If I'm reading that correctly, we should be able to assume that all files produced by CESM either (1) end with the date itself (I'm not sure whether any components actually do that?) or (2) end with [date].nc.

**Comment by billsacks** [comment]
@mvertens suggests - to be conservative about this, given how close we are to both the CESM and E3SM releases, and given that the short-term archiver has a history of bug whack-a-mole (fixing one bug just causes another to pop up somewhere else): Maybe for now we should leave the short-term archiver as is, and instead restrict case names. This seems like a reasonable solution to me.

For the case name restriction, I'd probably just say something like: you can't have the pattern `[0-9]-[0-9]` anywhere in the case name (that's a little more conservative than it needs to be, but is easier to document and easier to get right than a more permissive rule).

What do others think of this?

**Comment by mfdeakin-sandia** [comment]
I agree with your interpretation of the link, though @gold2718 's comment about history file name specifications does worry me. Could the short term archiver read those and give an error if they're set? Then we could just append the .nc to the regex.
As for changing things, I do have a bug fix for 10k dates I've been sitting on until I had time to fix the LDSTA test, so I don't think changing this now would be an issue?

**Comment by rljacob** [comment]
I'm in favor of not messing with the short term archiver if we can avoid it.
